src/runtime_src/xocl/api/clEnqueueReadBufferRect.cpp

```diff
@@ -145,30 +145,31 @@ clEnqueueReadBufferRect(cl_command_queue     command_queue ,
     xocl::xocl(*event)->queue(true /*wait*/);
   }
 
-  // Now the event is running, this should be hard_event and handle asynchronously
   auto device = xocl::xocl(command_queue)->get_device();
-  auto xdevice = device->get_xdevice();
   auto boh = xocl::xocl(buffer)->get_buffer_object_or_error(device);
-  void* host_ptr = xdevice->map(boh);
-  
-  size_t yit,zit;
-  for(zit=0;zit<region[2];zit++){
-    for(yit=0;yit<region[1];yit++){
-      size_t buffer_row_origin_in_bytes =
-        buffer_origin_in_bytes+
-        zit*buffer_slice_pitch+
-        yit*buffer_row_pitch;
+  cl_int errc = 0;
+
+  // This is memory map is bigger than we could make it when start offsets is not 0, 0, 0.
+  void *host_ptr = clEnqueueMapBuffer(command_queue, buffer, true, CL_MAP_READ,
```

Comment:
Just curious: is it possible to mix and match high-level call to OpenCL API inside lower-level `xocl` code?

Reply:
@keryell What do you mean ?  In this case there is no mixing. If you are asking about xocl::xocl, then it's just a static_cast from opencl domain to implementation domain.  `xocl `originally stood for xilinx opencl, before someone hijacked it for the userpf driver.

Reply:
Btw.  I have no clue about the validity of the changes in this pull request.  I trust @Ralender has the understanding, I don't think anyone is using this API and I am sure it is not tested in any of our test suites.  The code is leftover from the original sdaccel days, so thank you for fixing it!